File: src/BezierCurve.cpp
/*
  Copyright (C) 2009 Jon Macey

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/// @file BezierCurve.cpp
/// @brief basic BezierCurve using CoxDeBoor algorithm

#include "BezierCurve.h"
#include <iostream>
namespace ngl
{

void BezierCurve::createKnots() noexcept
{
	m_knots.clear();
	for(unsigned int i=0; i<m_numKnots; ++i)
	{
		m_knots.push_back( (i<(m_numKnots/2))  ? 0.0f : 1.0f);
	}
}


std::vector<Vec3> BezierCurve::getControlPoints() const noexcept
{
	return m_cp;
}
std::vector<Real> BezierCurve::getKnots() const noexcept
{
	return m_knots;
}



BezierCurve::BezierCurve( const std::vector<ngl::Vec3> &_p) noexcept : m_cp{_p}
{
	m_numCP=static_cast<unsigned int>(m_cp.size());
	m_degree=static_cast<unsigned int>(m_cp.size());
	m_order=m_degree+1;
	m_numKnots=m_numCP+m_order;
	createKnots();
}


BezierCurve::BezierCurve(const std::vector<Vec3> &_p, const std::vector<Real> &_k) noexcept : m_cp{_p}, m_knots{_k}
{
	m_numCP=static_cast<unsigned int>(_p.size());
	m_degree=static_cast<unsigned int>(_p.size());
	m_order=m_degree+1;
	m_numKnots=static_cast<unsigned int>(_k.size()); 
}


BezierCurve::~BezierCurve() noexcept
{
	m_cp.clear();
	m_knots.clear();
	if(m_vaoCurve!=nullptr && m_vaoPoints!=nullptr)
	{
		m_vaoCurve->unbind();
    m_vaoCurve->removeVAO();
		m_vaoPoints->unbind();
    m_vaoPoints->removeVAO();
	}
}


Real BezierCurve::coxDeBoor( Real _u,unsigned int _i, unsigned int _k, const std::vector <Real> &_knots ) const noexcept
{
	if(_k==1)
	{
		if( _knots[_i] <= _u && _u <= _knots[_i+1] )
		{
			return 1.0f;
		}
		return 0.0f;
	}
	Real Den1 = _knots[_i+_k-1] - _knots[_i];
	Real Den2 = _knots[_i+_k] - _knots[_i+1];
	Real Eq1=0;
	Real Eq2=0;
	if(Den1>0)
	{
		Eq1 = ((_u-_knots[_i]) / Den1) * coxDeBoor(_u,_i,_k-1,_knots);
	}
	if(Den2>0)
	{
		Eq2 = (_knots[_i+_k]-_u) / Den2 * coxDeBoor(_u,_i+1,_k-1,_knots);
	}
	return Eq1+Eq2;
}



void BezierCurve::drawControlPoints()const noexcept
{

  m_vaoPoints->bind();
  m_vaoPoints->setMode(GL_POINTS);
  m_vaoPoints->draw();
  m_vaoPoints->unbind();

}


void BezierCurve::drawHull()const noexcept
{
  m_vaoPoints->bind();
  m_vaoPoints->setMode(GL_LINE_STRIP);
  m_vaoPoints->draw();
  m_vaoPoints->unbind();
 }


void BezierCurve::draw() const noexcept
{
m_vaoCurve->bind();
m_vaoCurve->draw();
m_vaoCurve->unbind();
}


Vec3 BezierCurve::getPointOnCurve( Real _value  ) const noexcept
{
	Vec3 p;

	// sum the effect of all CV's on the curve at this point to
	// get the evaluated curve point
	//
  for(unsigned int i=0;i!=m_numCP;++i)
	{
		// calculate the effect of this point on the curve
		Real val = coxDeBoor(_value,i,m_degree /*was m_order */,m_knots);

		if(val>0.001f)
		{
			// sum effect of CV on this part of the curve
			p+=val*m_cp[i];
		}
	}

	return p;
}





void BezierCurve::addPoint( const Vec3 &_p    ) noexcept
{
	m_cp.push_back(_p);
	++m_numCP;
	++m_degree;
	m_order=m_degree+1;
	m_numKnots=m_numCP+m_order;
	createKnots();
	#ifdef DEBUG
    std::cout <<"Added "<<m_numCP<<" m_degree "<<m_degree<<" m_numKnots"<<m_numKnots<<" m_order "<<m_order<<'\n';
	#endif
}


void BezierCurve::addPoint( Real _x,  Real _y, Real _z) noexcept
{
	m_cp.push_back(Vec3(_x,_y,_z));
	++m_numCP;
	++m_degree;
	m_order=m_degree+1;
	m_numKnots=m_numCP+m_order;
	createKnots();

	#ifdef DEBUG
    std::cout <<"Added "<<m_numCP<<" m_degree "<<m_degree<<" m_numKnots"<<m_numKnots<<" m_order "<<m_order<<'\n';
	#endif
}

void BezierCurve::addKnot(Real _k) noexcept
{
	m_knots.push_back(_k);
	m_numKnots=m_numCP+m_order;
}

void BezierCurve::createVAO() noexcept
{
  if(m_vaoCurve!=nullptr && m_vaoPoints!=nullptr)
  {
    m_vaoCurve->unbind();
    m_vaoCurve->removeVAO();
    //delete m_vaoCurve;
    m_vaoPoints->unbind();
    m_vaoPoints->removeVAO();
    //delete m_vaoPoints;
  }

  m_vaoPoints=ngl::VAOFactory::createVAO("simpleVAO",GL_POINTS);
  m_vaoPoints->bind();
  size_t size=m_cp.size();
  std::vector <Vec3> points(size);
  for(size_t i=0;i<size;++i)
  {
    points[i].set(m_cp[i]);
  }
  m_vaoPoints->setData(SimpleVAO::VertexData(m_numCP*sizeof(Vec3),points[0].m_x));
  m_vaoPoints->setNumIndices(m_numCP);
  m_vaoPoints->setVertexAttributePointer(0,3,GL_FLOAT,0,0);
  m_vaoPoints->unbind();


  m_vaoCurve=ngl::VAOFactory::createVAO("simpleVAO",GL_LINE_STRIP);
  m_vaoCurve->bind();

  std::vector <Vec3> lines(m_lod);
  for(unsigned int i=0;i!=m_lod;++i)
  {
    Real t  = m_knots[m_numKnots-1] * i / static_cast<Real>(m_lod-1);

    if(i==m_lod-1)
    {
      t-=0.001f;
    }
    lines[i].set(getPointOnCurve(t));
  }
  m_vaoCurve->setData(SimpleVAO::VertexData(m_lod*sizeof(Vec3),lines[0].m_x));
  m_vaoCurve->setNumIndices(m_lod);
  m_vaoCurve->setVertexAttributePointer(0,3,GL_FLOAT,0,0);
  m_vaoCurve->unbind();

}



} // end ngl namespace

File: py-ngl/src/ngl/bezier_curve.py
from .vec3 import Vec3


class BezierCurve:
    """A Bezier curve class."""

    def __init__(
        self, control_points: list[Vec3] = None, knots: list[float] = None
    ) -> None:
        self._cp = control_points if control_points is not None else []
        self._knots = knots if knots is not None else []
        self._degree = 0
        self._order = 0
        self._num_cp = 0
        self._num_knots = 0
        if self._cp:
            self._num_cp = len(self._cp)
            self._degree = self._num_cp
            self._order = self._degree + 1
            if not self._knots:
                self.create_knots()
            self._num_knots = len(self._knots)

    @property
    def control_points(self) -> list[Vec3]:
        return self._cp

    @property
    def knots(self) -> list[float]:
        return self._knots

    def add_point(self, x: float | Vec3, y: float = None, z: float = None) -> None:
        if isinstance(x, Vec3):
            self._cp.append(x)
        else:
            self._cp.append(Vec3(x, y, z))
        self._num_cp += 1
        self._degree = self._num_cp
        self._order = self._degree + 1
        self.create_knots()

    def add_knot(self, k: float) -> None:
        self._knots.append(k)
        self._num_knots = len(self._knots)

    def create_knots(self) -> None:
        self._num_knots = self._num_cp + self._order
        self._knots = [0.0] * (self._num_knots // 2) + [1.0] * (
            self._num_knots - (self._num_knots // 2)
        )

    def get_point_on_curve(self, u: float) -> Vec3:
        p = Vec3()
        for i in range(self._num_cp):
            val = self.cox_de_boor(u, i, self._degree, self._knots)
            if val > 0.001:
                p += self._cp[i] * val
        return p

    def cox_de_boor(self, u: float, i: int, k: int, knots: list[float]) -> float:
        if k == 1:
            return 1.0 if knots[i] <= u <= knots[i + 1] else 0.0

        den1 = knots[i + k - 1] - knots[i]
        den2 = knots[i + k] - knots[i + 1]

        eq1 = 0.0
        if den1 > 0:
            eq1 = ((u - knots[i]) / den1) * self.cox_de_boor(u, i, k - 1, knots)

        eq2 = 0.0
        if den2 > 0:
            eq2 = ((knots[i + k] - u) / den2) * self.cox_de_boor(u, i + 1, k - 1, knots)

        return eq1 + eq2
